discrete_math/representations_of_numbers/representation.hpp

```diff
@@ -0,0 +1,147 @@
+#include <iostream>
+#include <concepts>
+#include <variant>
+#include <expected>
+#include <variant>
+#include <vector>
+#include <string>
+
+namespace representation_namespace {
+
+    const int NUMBERS = 48;
+    const int CHARS = 55;
+
+    template <std::integral I>
+    struct fraction {
+        I numerator{ 0 };
+        I denominator{ 1 };
+
+        fraction(I num = 0, I den = 1) : numerator(num), denominator(den) {}
+    };
+
+    template <std::uint8_t base>
+    struct expansion {
+        static constexpr std::uint8_t form = base;
+    };
+
+    class representation {
+    public:
+
+        template <std::integral I>
+        fraction<I> create_fraction(I, I);
+
+        template <std::uint8_t base>
+        expansion<base> create_expansion(int);
+
+        template <std::integral I>
+        std::string plus_or_minus(fraction<I>);
+
+        template <std::integral I, std::uint8_t base>
+        std::string integral_number(I, I, expansion<base>);
+
+        template <std::integral I, std::uint8_t base>
+        std::string expansion_in_base(fraction<I>, expansion<base>);
+
+        template <std::integral I>
+        std::vector<I> set_of_rest(std::string&, I, I, I, int);
+
+        template <std::integral I>
+        std::string periodic(std::vector<I>, std::string, I);
+
+        template <std::integral I, std::uint8_t base>
+        std::string fractional_number(I, I, expansion<base>);
+    };
+
+    template <std::integral I>
+    fraction<I> representation::create_fraction(I numerator, I denominator) {
+        return fraction<I>(numerator, denominator);
+    }
+
+    template <std::uint8_t base>
+    expansion<base> representation::create_expansion(int) {
+        return expansion<base>();
+    }
+
+    template <std::integral I>
+    std::string representation::plus_or_minus(fraction<I> my_fraction) {
+        if ((my_fraction.numerator > 0 && my_fraction.denominator > 0) || (my_fraction.numerator < 0 && my_fraction.denominator < 0)) {
+            return " ";
+        }
+        else {
+            return "-";
+        }
+    }

```

Comment:
you can use convention that numerator is an integer and denominator is natural number. Then sign of fraction is sign of the numerator